### PR TITLE
fix: use type `std::size_t` for index in `Collection` element accessors and `size()`

### LINKED ISSUE #406: type inconsistency between `Collection::size()` and index for const object accessors
- PODIO version: `master`

There is a difference between the type `size_t` returned by `size()` from a Collection and the type of the index used in the const object accessors, `unsigned int`:

https://github.com/AIDASoft/podio/blob/f4c9219ddde59b1efc73db3fd9c25139222a7b3c/python/templates/Collection.h.jinja2#L77

vs.

https://github.com/AIDASoft/podio/blob/f4c9219ddde59b1efc73db3fd9c25139222a7b3c/python/templates/Collection.h.jinja2#L94-L101

Is there a particular reason for this difference?

_cf._ `std::vector`, where we have (ignoring C++20 changes for clarity):
```cpp
size_type size();
reference at( size_type pos );
```
in other words, `std::vector::at` takes the same type that `std::vector::size()` returns.

### PR DESCRIPTION
**Pull request by c-dilks**
BEGINRELEASENOTES
- fix type inconsistency between `Collection::size()` and index for const object accessors

ENDRELEASENOTES

fix #406

### CODE REVIEW

**Comment by c-dilks** [comment]
Looks like the `key4hep` CI failure is also happening on `master`.

I can try to make `VectorMembers` and `OneToManyRelations` use `size_t` here too, but I may need help making sure everything that needs to change is changed.

**Comment by tmadlener** [comment]
> Looks like the `key4hep` CI failure is also happening on `master`.

Yes that is a slightly annoying but known issue, see #403.

File: python/templates/Data.h.jinja2 [hunk 1]
```diff
@@ -19,8 +19,8 @@ public:
 {% endfor %}
 
 {% for member in VectorMembers + OneToManyRelations %}
-  unsigned int {{ member.name }}_begin{};
-  unsigned int {{ member.name }}_end{};
+  size_t {{ member.name }}_begin{};
```
Comment (tmadlener) [inline python/templates/Data.h.jinja2:22]: This would be a breaking change for generated datamodels. Is this strictly necessary?

**Comment by c-dilks** [comment]
All right, I think that's all of them. I didn't bother to change the internal `for` loops, e.g., https://github.com/AIDASoft/podio/blob/f4c9219ddde59b1efc73db3fd9c25139222a7b3c/python/templates/macros/implementations.jinja2#L198
since that may be a bit overkill; my main concern was from the user-facing side of the generated C++ code. 

I didn't touch any of the `SIOBlock` code, but at first glance I didn't find any user-facing `unsigned` types to worry about.

Do we need to add `size_t` anywhere in this file?
https://github.com/AIDASoft/podio/blob/f4c9219ddde59b1efc73db3fd9c25139222a7b3c/python/podio/generator_utils.py#L41-L54

**Comment by tmadlener** [comment]
The `BUILTIN_TYPES` effectively just determine which types can be used in the yaml files to define a datamodel, so I think we don't need `size_t` in there. At least until someone actually requests it.

Thanks again for going through all the bits and pieces here. Waiting for CI to succeed and then I think this can be merged.

**Comment by jmcarcell** [comment]
Maybe `std::size_t` instead of `size_t`? I don't have an opinion on that

**Comment by tmadlener** [comment]
Good point. I don't have a strong opinion on it either. Strictly speaking `#include <cstddef>` and `std::size_t` would probably be the most standard conformant way to do this. For all practical purposes it probably doesn't matter too much, but since we are already here we might as well go ahead and use `std::size_t` and add the necessary include to the `Collection.h.jinja2` template.

**Comment by c-dilks** [comment]
> Good point. I don't have a strong opinion on it either. Strictly speaking `#include <cstddef>` and `std::size_t` would probably be the most standard conformant way to do this. For all practical purposes it probably doesn't matter too much, but since we are already here we might as well go ahead and use `std::size_t` and add the necessary include to the `Collection.h.jinja2` template.

Done in f9880d6.

Do we need `#include<cstddef>` in `Object.h.jinja2` and `MutableObject.h.jinja2` too, or anywhere else?

**Pull request merged**